boa/src/builtins/date/mod.rs

```diff
@@ -0,0 +1,1421 @@
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    builtins::{
+        function::{make_builtin_fn, make_constructor_fn},
+        object::ObjectData,
+        ResultValue, Value,
+    },
+    exec::PreferredType,
+    BoaProfiler, Interpreter,
+};
+use chrono::{prelude::*, Duration, LocalResult};
+use gc::{unsafe_empty_trace, Finalize, Trace};
+use std::fmt::Display;
+
+const NANOS_IN_MS: f64 = 1_000_000f64;
+
+#[inline]
+fn is_zero_or_normal_opt(value: Option<f64>) -> bool {
+    value
+        .map(|value| value == 0f64 || value.is_normal())
+        .unwrap_or(true)
+}
+
+#[inline]
+fn ignore_ambiguity<T>(result: LocalResult<T>) -> Option<T> {
+    match result {
+        LocalResult::Ambiguous(v, _) => Some(v),
+        LocalResult::Single(v) => Some(v),
+        LocalResult::None => None,
+    }
+}
+
+/// Some JS functions allow completely invalid dates, and the runtime is expected to make sense of this. This function
+/// constrains a date to correct values.
+fn fix_date(year: &mut i32, month: &mut i32, day: &mut i32) {
+    #[inline]
+    fn num_days_in(year: i32, month: u32) -> i32 {
+        let month = month + 1; // zero-based for calculations
+        NaiveDate::from_ymd(
+            match month {
+                12 => year + 1,
+                _ => year,
+            },
+            match month {
+                12 => 1,
+                _ => month + 1,
+            },
+            1,
+        )
+        .signed_duration_since(NaiveDate::from_ymd(year, month, 1))
+        .num_days() as i32
+    }
+
+    #[inline]
+    fn fix_month(year: &mut i32, month: &mut i32) {
+        *year += *month / 12;
+        *month = if *month < 0 {
+            *year -= 1;
+            11 + (*month + 1) % 12
+        } else {
+            *month % 12
+        }
+    }
+
+    #[inline]
+    fn fix_day(year: &mut i32, month: &mut i32, day: &mut i32) {
```

Comment:
Maybe this would be more efficient by passing the values and returning a result, since they should be easily copied in the stack. Maybe this would be easier by passing a custom `Date` structure around, with these three fields, and having the "fix" functions as methods.

Reply:
I have solved this by making `Date` the "custom structure." There's a new `set_components` method that all the other setters call.